### PR TITLE
Moves SearchMetadata component from App to Results

### PR DESCRIPTION
**Pull request by JPrevost**
## Why are these changes being introduced:

* The SearchMetadata is only used as part of the rendering logic for
  Results. As such, it is easier to work with it as a child Component
  rather than a sibling.

## Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/DISCO-195

## How does this address that need:

* This moves the SearchMetadata component to be a child of Results

## Document any side effects to this change:

* I've removed some additional data elements from App that I noticed
  were unused when moving the SearchMetadata component
* I've removed the Related component from App and did not add it any
  place else. I feel like it'll be part of Record, but until we decide
  for sure I'd prefer not to include it at all. I've left Help and About
  as those seem like App level components and will likely have clear
  purpose in the near future.
* I've included some pagination metadata that didn't land with the
  Pagination work as there wasn't a clear place to display it and I
  knew this work was upcoming.

#### Developer

- [x] All new ENV is documented in README
- [x] All new ENV has been added to Heroku Pipeline, Staging and Prod
- [x] ANDI or Wave has been run in accordance to
      [our guide](https://mitlibraries.github.io/guides/basics/a11y.html) and
      all issues introduced by these changes have been resolved or opened as new
      issues (link to those issues in the Pull Request details above)
- [x] Stakeholder approval has been confirmed (or is not needed)

#### Code Reviewer

- [x] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [x] There are appropriate tests covering any new functionality
- [x] The documentation has been updated or is unnecessary
- [x] The changes have been verified
- [x] New dependencies are appropriate or there were no changes

#### Includes new or updated dependencies

NO

### CODE REVIEW

File: src/components/SearchMetadata.vue [hunk 1]
```diff
@@ -11,23 +13,17 @@ export default {
...
-    panelHeading: function () {
-      if (this.status.ready == true) {
-        return "Results summary: " + this.hits + " results";
-      } else if (this.status.loading == true) {
-        return "Loading results...";
-      }
-      return "Search Metadata:";
+  methods: {
+    per_page() {
+      return 20;
```
Comment (matt-bernhardt) [inline src/components/SearchMetadata.vue:20]: We already store this value in Results.vue (line 52 or 55 depending on which branch) - could it be passed into this SFC, rather than re-computed?

File: src/components/SearchMetadata.vue [hunk 2]
```diff
@@ -11,23 +13,17 @@ export default {
...
     },
-    panelStatus: function () {
-      if (this.status.ready == true) {
-        return "panel-success";
-      }
-      return "panel-info";
+    start() {
+      return this.$route.query.page * this.per_page() - this.per_page() + 1;
```
Comment (matt-bernhardt) [inline src/components/SearchMetadata.vue:23]: This is mathematically correct, and I don't think we _need_ to make it more efficient - but it can be done without calling per_page() twice if the expression is `this.per_page() * ( this.$route.query.page - 1 ) + 1`

File: src/views/Results.vue [hunk 3]
```diff
@@ -14,8 +14,9 @@
       <Facet v-for="facet in facets" v-bind:facet="facet" v-bind:key="facet" />
     </div>
     <div v-bind:class="contentClass" v-if="this.status.loading == false">
+      <SearchMetadata v-model:searchterm="searchterm" v-model:hits="hits" />
```
Comment (matt-bernhardt) [inline src/views/Results.vue:17]: I don't object to this, but want to make sure we're choosing it intentionally. This would put the SearchMetadata display alongside any chosen facets, rather than above them (which is where the other status-report-like UI elements would appear).

File: src/views/Results.vue [hunk 4]
```diff
@@ -64,24 +67,21 @@ export default {
...
-      return this.hits > 0;
-    },
     showSidebar: function () {
       return this.facets.length ? true : false;
     },
   },
   methods: {
     async searchTimdex() {
       this.status.loading = true;
+      this.searchterm = this.$route.query.q;
```
Comment (matt-bernhardt) [inline src/views/Results.vue:77]: Is there a reason that we retrieve and store the search query twice? We store it in `this.searchterm` here, but a few lines lower we also grab it via `let query = qs.stringify(this.$route.query)`. Is there a reason for not stringifying this version, and using the `.q` portion here, versus the other choices below?

File: tests/unit/searchmetadata.spec.js [hunk 5]
```diff
@@ -3,28 +3,97 @@ import SearchMetadata from "@/components/SearchMetadata.vue";
...
+      },
+      props: { hits, searchterm },
     });
-    expect(wrapper.text()).toMatch("Results summary:");
-    expect(wrapper.text()).toMatch(query);
+
+    expect(wrapper.text()).toMatch("Search Results");
+    expect(wrapper.text()).toMatch(searchterm);
     expect(wrapper.text()).toMatch(String(hits));
+    expect(wrapper.text()).toMatch("Viewing records 1 to 20 of " + hits);
```
Comment (matt-bernhardt) [inline tests/unit/searchmetadata.spec.js:28]: These two tests seem to be basically the same thing? One just looking for the string of the hit count, while the other tests for the specific calculated phrase. The rest of the tests in this file just look for the calculated phrase. My biggest preference would be to consistent in the tests, although I'm also not sure if this is the sort of UI-specific thing that we shouldn't be testing for?

I will say, though, that I'm still getting used to the mindset of appropriately-thorough testing - and I can see some logic in making sure that the composition of the phrase is correct.

**Review by matt-bernhardt — COMMENTED** [review]
I've left some questions on specific lines, but also had some questions that seem less easy to isolate - so I'll ask them here with some screenshots or links.

1. Are we intending to make the `page` querystring required? If I try going to just http://localhost:8080/results?q=default%20value%20of%20number I get some NaN values in the SearchMetadata component:
Screen Shot 2020-11-23 at 2 04 15 PM

2. For cases where there's an error or API outage, I'm seeing both the "Something went wrong" and the SearchMetadata component. Is that expected, or should only the error be shown if that happens?

(with the TIMDEX_API value set incorrectly, I get this)
Screen Shot 2020-11-23 at 2 22 00 PM

I don't think I'm strongly opposed enough to any of these right now that I'd ask that they be changed - but I'd like to understand why?

File: src/components/SearchMetadata.vue (hunk 1 above)
Comment (JPrevost) [reply src/components/SearchMetadata.vue:20]: yeah the whole `per_page` thing is a mess. It needs to be set in a single place for sure but I was nervous having it be ENV because it doesn't "change" what we pull back yet. I think I'll open a ticket to address both moving all of the random places it's set in the app _and_ implementing the TIMDEX call to allow changing it for real.

File: src/components/SearchMetadata.vue (hunk 2 above)
Comment (JPrevost) [reply src/components/SearchMetadata.vue:23]: #team_no_math. But also I'll look closer at that expression and see if I feel like it makes it clearer or less clear what we're doing. 👍🏻

File: src/views/Results.vue (hunk 4 above)
Comment (JPrevost) [reply src/views/Results.vue:77]: the `.q` is pulling in just one parameter from the query string for display of the serachterm. The stringified thing here pulls in all of the current and future parameters (which will include facets). I'm sure we could make one less call to `this.$route.query` but I'm not sure it would be more or less confusing. I'll take a look.

File: tests/unit/searchmetadata.spec.js (hunk 5 above)
Comment (JPrevost) [reply tests/unit/searchmetadata.spec.js:28]: yeah the composition of the phrase is really important as the logic is a bit messy and involves math and edge cases we need to handle. I don't think just because we want to watch different values of that one phrase across a bunch of variants we need to test each step of the way like we do in that first test. The first test could be split up into separate tests but I don't personally see much value of running the same inputs in and testing once for the general state and once for the `viewing records...` I feel like the tests are fine here but am definitely willing to talk about this further to better understand your stance on this.

File: src/views/Results.vue (hunk 3 above)
Comment (JPrevost) [reply src/views/Results.vue:17]: I agree with what you are saying in that it should be intentional to determine where things display, but I'm choosing for now to wait until we have real facets displaying to make this decision. If Adam lands his code first, we can definitely address this. For now I'm pretending that non-displaying component doesn't exist as I'm confident it has changed way to much for me to make assumptions about here.

**Comment by JPrevost** [comment]
@matt-bernhardt I'll add a new test and handle the lack of passing `page` in via the route query. Good catch.

**Comment by JPrevost** [comment]
As for displaying partial metadata and an error. I'm going to let UX decide what they want going on here. I think both make sense to display but definitely agree that the current order / nesting of them is wonky.

File: src/views/Results.vue (hunk 4 above)
Comment (JPrevost) [reply src/views/Results.vue:77]: @matt-bernhardt we need to grab `query.q` here to allow us to have that set regardless of whether the timdex call succeeds (we use it elsewhere regardless of success). We grab the full query as a convenient way to populate all of the data we'll need for the timdex call.

I'm not sure if you are suggesting a change or just trying to understand the logic here. You are definitely correct that `q` is a subset of the full `query` but as we are using them for different purposes I'm not sure it would be clearer to change this. (Also, this changeset just moved this logic outside of the condition as we needed it regardless of success).

File: tests/unit/searchmetadata.spec.js (hunk 5 above)
Comment (JPrevost) [reply tests/unit/searchmetadata.spec.js:28]: @matt-bernhardt I've been thinking about this a bit and I think what we really need to test is the the methods `start()`, `end()`, `page()`, and `per_page()` return the proper results. However, I haven't seen a way to do that in Vue. It seems to really want you to just check that the UI is in the state you expect rather than testing the individual methods.

If there are further things you'd like to discuss about this let me know. Otherwise I'm inclined to resolve this comment.

File: src/components/SearchMetadata.vue (hunk 2 above)
Comment (JPrevost) [reply src/components/SearchMetadata.vue:23]: Yeah I like this better. I'm going to make this change. Thanks for the suggestion.

File: src/components/SearchMetadata.vue (hunk 1 above)
Comment (matt-bernhardt) [reply src/components/SearchMetadata.vue:20]: Yeah, if that's better handled in a ticket for focused cleanup, that's fine with me - particularly if the work extends beyond the bounds of this PR (which it seems like it does).

File: src/views/Results.vue (hunk 4 above)
Comment (matt-bernhardt) [reply src/views/Results.vue:77]: I'm happy to just understand the reasons for doing it this way - I noted that this was now coming before the if statement, which makes sense as you say. Maybe I need to do some more digging around this operation myself and build my understanding of what's happening with `query.q` versus `query`.

Long story short - if you've looked at this again, in light of my questions, and still think this is the way it needs to be, then I'm not going to object. If it still doesn't make sense to me later we can have a more focused conversation then.

Comment (JPrevost) [reply src/views/Results.vue:77]: Oh gotcha. `query` is the full url query string. So it includes things like `q=whatever you searched for`, `page=1`, `facet_whatever=cheese`.

`query.q` is just pulling out our search term which we've named `q`.

This confusion is why I've started to move us to internally calling the searched for phrase `searchterm` instead of `query` because `query` encompasses a lot more in a vue route and it was making it hard to keep track of.

Comment (matt-bernhardt) [reply src/views/Results.vue:77]: Yeah, that makes sense.

File: src/components/SearchMetadata.vue (hunk 1 above)
Comment (JPrevost) [reply src/components/SearchMetadata.vue:20]: https://mitlibraries.atlassian.net/browse/DISCO-209

File: src/views/Results.vue (hunk 3 above)
Comment (matt-bernhardt) [reply src/views/Results.vue:17]: Resolving this...

**Review by matt-bernhardt — APPROVED** [review]
This all looks good to me now - thanks for clarifying and updating in various ways.

:shipit:
